yarn-project/ethereum/src/gas_utils.ts

```diff
@@ -0,0 +1,216 @@
+    );
+
+    return finalGasPrice;
+  }
+
+  /**
+   * Estimates gas with retries and adds buffer
+   */
+  public estimateGas<T extends (...args: any[]) => Promise<bigint>>(estimator: T): Promise<bigint> {
+    return retry(
+      async () => {
+        const gas = await estimator();
+        return gas + (gas * this.gasConfig.bufferPercentage) / 100n;
+      },
+      'gas estimation',
+      makeBackoff([1, 2, 3]),
+      this.logger,
+      false,
+    );
+  }
+
+  /**
+   * Monitors a transaction and attempts to speed it up if it gets stuck
+   * @param txHash - The hash of the transaction to monitor
+   * @param walletClient - The wallet client for sending replacement tx
+   * @param originalTx - The original transaction data for replacement
+   * @param config - Monitor configuration
+   */
+  public async monitorTransaction(
+    txHash: Hex,
+    walletClient: WalletClient,
+    originalTx: {
+      to: Address;
+      data: Hex;
+      nonce: number;
+      gasLimit: bigint;
+      maxFeePerGas: bigint;
+    },
+    txMonitorConfig?: TransactionMonitorConfig,
+  ): Promise<TransactionReceipt> {
+    const config = { ...this.monitorConfig, ...txMonitorConfig };
+    let attempts = 0;
+    let lastSeen = Date.now();
+    let currentTxHash = txHash;
+
+    // Flag to use for first run to avoid logging errors.
+    let firstRun = true;
+
+    while (true) {
+      try {
+        // Check transaction status
+        const receipt = await this.publicClient.getTransactionReceipt({ hash: currentTxHash });
+        if (receipt) {
+          this.logger?.info(`Transaction ${currentTxHash} confirmed`);
+          return receipt;
+        }
+
+        if (firstRun) {
+          firstRun = false;
+          // Wait a second for tx to be broadcast
+          await sleep(1000);
+        }
+
+        // Check if transaction is pending
+        const tx = await this.publicClient.getTransaction({ hash: currentTxHash });
+        this.logger?.info(`Transaction ${currentTxHash} pending`);
+        if (tx) {
+          lastSeen = Date.now();
+          await new Promise(resolve => setTimeout(resolve, config.checkIntervalMs));
+          continue;
+        }
+
+        // Transaction not found and enough time has passed - might be stuck
+        if (Date.now() - lastSeen > config.stallTimeMs && attempts < config.maxAttempts) {
+          attempts++;
+          const newGasPrice = (originalTx.maxFeePerGas * (100n + config.gasPriceIncrease)) / 100n;
+
+          this.logger?.info(
+            `Transaction ${currentTxHash} appears stuck. Attempting speed-up ${attempts}/${config.maxAttempts} ` +
+              `with new gas price ${formatGwei(newGasPrice)} gwei`,
+          );
+
+          // Send replacement transaction with higher gas price
+          const account = await walletClient.getAddresses().then(addresses => addresses[0]);
```

Comment:
I understand that if you use a `WalletClient<HttpTransport, Chain, Account>`, which you get by [hoisting the account](https://viem.sh/docs/clients/wallet#optional-hoist-the-account), you don't need this (and don't risk accidentally sending a tx as a different account!).